java-checks/src/main/java/org/sonar/java/checks/serialization/SerializableObjectInSessionCheck.java

```diff
@@ -44,8 +47,45 @@
   protected void onMethodInvocationFound(MethodInvocationTree mit) {
     ExpressionTree argument = mit.arguments().get(1);
     Type type = argument.symbolType();
-    if (!type.isPrimitive() && !type.isSubtypeOf("java.io.Serializable")) {
-      reportIssue(argument, "Make \"" + type + "\" serializable or don't store it in the session.");
+    if (!isSerializable(type)) {
+      String andParameters = isParametrized(type) ? " and its parameters" : "";
+      reportIssue(argument, "Make \"" + type + "\"" + andParameters + " serializable or don't store it in the session.");
     }
   }
+
+  private boolean isSerializable(Type type) {
+    if (type.isPrimitive()) {
+      return true;
+    }
+    if (isSerializableArray(type)) {
+      return true;
+    }
+    if (isParametrized(type)) {
+      return isSerializableParametrized((ParametrizedTypeJavaType) type);
+    }
+    return type.isSubtypeOf("java.io.Serializable");
+  }
+
+  private boolean isSerializableArray(Type type) {
+    if (type instanceof ArrayJavaType) {
+      ArrayJavaType arrayJavaType = (ArrayJavaType) type;
+      return isSerializable(arrayJavaType.elementType());
+    }
+    return false;
+  }
+
+  private boolean isParametrized(Type type) {
+    return type instanceof ParametrizedTypeJavaType;
+  }
+
+  private boolean isSerializableParametrized(ParametrizedTypeJavaType type) {
+    // note: this is assuming that custom implementors of Collection
+    // have the good sense to make it serializable just like all implementations in the JDK
+    return (type.isSubtypeOf("java.io.Serializable") || type.isSubtypeOf("java.util.Collection"))
+      && type.typeParameters().stream().allMatch(t -> {
+      JavaType javaType = type.substitution(t);
+      assert javaType != null;
```

Comment:
Please remove the `assert` statement and replace the whole lambda expression by:
```
t -> isSerializable(type.substitution(t));
```

The situation where we are not able to retrieve the substitution for one of its own type parameter should never occur. However, if it occurs, we don't want the rule to fail the analysis with an assertion error, we want it to fail badly. It means that we are facing a problematic situation and we want a feedback from an afraid user on it. It is the responsibility of the semantic analysis to fail if it encounter an impossible situation, not the rules.

From a coverage point of view, you also won't be able to cover such cases, as you can't build it.